File: coconut2d-x/layers/CNLayer.cpp
//
//  CNLayer.cpp
//  Coconut2D-X
//
//  Created by Moky on 12-11-22.
//  Copyright 2012 Slanissue.com. All rights reserved.
//

#include "CNLayer.h"

#include "../CNDirector.h"
#include "../CNNode.h"
#include "../sprites/CNSprite.h"

NS_CN_BEGIN

USING_NS_CC;

CNLayer::~CNLayer()
{
	CNLog("~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ dealloc");
}

bool CNLayer::init(void)
{
	if (!CCLayer::init())
	{
		CNLog("error");
		return false;
	}
	setAnchorPoint(CCPointZero);
	setIsRelativeAnchorPoint(true);
	setPosition(CCPointZero);
	return true;
}

bool CNLayer::initWithFile(const char * pszFilename)
{
	if (!init())
	{
		CNLog("error");
		return false;
	}
	CCNode * pBackground = CNSprite::spriteWithFile(pszFilename);
	pBackground->setAnchorPoint(CCPointZero);
	pBackground->setPosition(CCPointZero);
	return initWithNode(pBackground);
}

bool CNLayer::initWithNode(CCNode * pNode)
{
	if (!init())
	{
		CNLog("error");
		return false;
	}
	CCSize tSize = pNode->getContentSize();
	CCPoint tAnchor = pNode->getAnchorPointInPoints();
	CCPoint tOrigin = ccpSub(pNode->getPosition(), tAnchor);
	
	CCRect tFrame = CCRectZero;
	tFrame.size = tSize;
	tFrame.origin = tOrigin;
	if (!initWithFrame(tFrame))
	{
		CNLog("error");
		return false;
	}
	pNode->setAnchorPoint(CCPointZero);
	pNode->setPosition(CCPointZero);
	addChild(pNode);
	return true;
}

bool CNLayer::initWithFrame(const CCRect & tFrame)
{
	if (!init())
	{
		CNLog("error");
		return false;
	}
	setContentSize(tFrame.size);
	setPosition(tFrame.origin);
	return true;
}

bool CNLayer::initWithDictionary(CCDictionary * pDict)
{
	if (!init())
	{
		CNLog("error");
		return false;
	}
	return setAttributes(pDict);
}

void CNLayer::focus(const CCPoint & tPoint)
{
	CCSize tWinSize = CNDirector::sharedDirector()->getWinSize();
	CCPoint tCenter = ccpMult(ccpFromSize(tWinSize), 0.5f);
	CCPoint tAnchor = getAnchorPointInPoints();
	CCPoint tDistance = ccpSub(tAnchor, tPoint);
	CCPoint tNewPosition = ccpAdd(tCenter, tDistance);
	setPosition(tNewPosition);
}

CCPoint CNLayer::focus(void)
{
	CCSize tWinSize = CNDirector::sharedDirector()->getWinSize();
	CCPoint tCenter = ccpMult(ccpFromSize(tWinSize), 0.5f);
	return ccpSub(tCenter, getPosition());
}

NS_CN_END

File: coconut2d-ios/layers/CNLayer.m
//
//  CNLayer.m
//  Coconut2D
//
//  Created by Moky on 12-7-6.
//  Copyright 2012 Slanissue.com. All rights reserved.
//

#import "CNLayer.h"

#import "cnMacros.h"
#import "CNDirector.h"
#import "CNNode.h"
#import "CNSprite.h"


@implementation CNLayer

+ (id) layerWithFile:(NSString *)filename
{
	return [[[self alloc] initWithFile:filename] autorelease];
}

+ (id) layerWithNode:(CCNode *)background
{
	return [[[self alloc] initWithNode:background] autorelease];
}

+ (id) layerWithFrame:(CGRect)frame
{
	return [[[self alloc] initWithFrame:frame] autorelease];
}

+ (id) layerWithDictionary:(NSDictionary *)dict
{
	return [[[self alloc] initWithDictionary:dict] autorelease];
}

- (id) init
{
	if ((self = [super init]))
	{
		self.contentSize = [[CNDirector sharedDirector] winSize];
	}
	return self;
}

- (id) initWithDictionary:(NSDictionary *)dict
{
	if ((self = [self init]))
	{
		[self setAttributes:dict];
	}
	return self;
}

- (id) initWithFrame:(CGRect)frame
{
	if ((self = [self init]))
	{
		self.isRelativeAnchorPoint = YES;
		self.contentSize = frame.size;
		self.position = frame.origin;
	}
	return self;
}

- (id) initWithNode:(CCNode *)background
{
	CGSize size = background.contentSize;
	CGPoint anchor = ccpCompMult(ccpFromSize(size), background.anchorPoint);
	CGPoint origin = ccpSub(background.position, anchor);
	
	CGRect frame = CGRectZero;
	frame.size = size;
	frame.origin = origin;
	
	if ((self = [self initWithFrame:frame]))
	{
		background.anchorPoint = CGPointZero;
		background.position = CGPointZero;
		
		[self addChild:background];
	}
	return self;
}

- (id) initWithFile:(NSString *)filename
{
	CCNode * background = [[CNSprite alloc] initWithFile:filename];
	background.anchorPoint = CGPointZero;
	background.position = CGPointZero;
	self = [self initWithNode:background];
	[background release];
	return self;
}

- (void) dealloc
{
	CNLog(@"~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ %@ dealloc", [self class]);
	[super dealloc];
}


#pragma mark -


- (void) addChild:(CCNode *)child
{
	if (child == nil)
	{
		return ;
	}
	[super addChild:child];
}

- (void) addChild:(CCNode *)child z:(NSInteger)z
{
	if (child == nil)
	{
		return ;
	}
	[super addChild:child z:z];
}

- (void) addChild:(CCNode *)child z:(NSInteger)z tag:(NSInteger)tag
{
	if (child == nil || child.parent != nil)
	{
		return ;
	}
	[super addChild:child z:z tag:tag];
}

#pragma mark -

- (void) focus:(CGPoint)point
{
	CGSize winSize = [[CNDirector sharedDirector] winSize];
	CGPoint center = ccpMult(ccpFromSize(winSize), 0.5f);
	CGPoint anchor = ccpCompMult(ccpFromSize(self.contentSize), self.anchorPoint);
	CGPoint distance = ccpSub(anchor, point);
	CGPoint newPosition = ccpAdd(center, distance);
	self.position = newPosition;
}

- (CGPoint) focus
{
	CGSize winSize = [[CNDirector sharedDirector] winSize];
	CGPoint center = ccpMult(ccpFromSize(winSize), 0.5f);
	return ccpSub(center, self.position);
}

@end
